### ISSUE TITLE
Hiccups that where not always there

### ISSUE THREAD

**Issue by jgeisler0303** [badaix/snapcast#542]
Hi,
I am running a snapserver on a Banana Pi M1 with Armbian since a while now. In the beginning everything was beautiful but then it started to happen that the sound had "hiccups". I am feeding snapserver with mpd but the same problem occurs using mplayer. I am not really sure if there are very short durations of silence or if a short bit is skipped. It's always in the order of one beat or less. when it happens, it occurs about every 15 to 30 seconds. The hiccups seems to coincide with a `(onResync) onResync (default)` message on the server side with values around some 10ms and some times up to 200ms. On the client side the first two `Chunk` values are usually single digit but right before the hiccup they reach up to -2000. Then follows a `pShortBuffer->full() && (abs(shortMedian_) > 5)` line and then several lines like `Sleep -59, age: -59, bufferDuration` or  `sleep < -bufferDuration/2: -59 < -15, sleep`.

Often I used to be able to resolve the problem by restarting the client or the server. Later, only rebooting the whole Pi helped and now the problem seems to be persistent. In the mean time I have changed quite some things. Armbian was updated at least twice and snapserver /client as well. I am running the latest master branch self-compiled now.
Is there anything I can check that might have changed that causes the problem? E.g.

- NTP not running correctly
- Different time zones
- The clock of the CPU not being calibrated correctly / having some drift
- The CPU governor being more aggressive

Or is there maybe some regression in the snap synchronization algorithm?

I'd be willing to try some things to get this working again. E.g. some changes to the source code an recompile to test.

**Comment by badaix** [comment]
onResync happens when the server cannot read enough data from the pipe. Do you have the correct sample rate configured? If the source rate is 44.1kHz and the Snapserver rate is 48kHz, the server will read faster.
Did this start after upgrading to a newer Snapcast version?
There is currently some client side optimization on the jitter and synchronization going on in develop branch, but onResync is a server issue.
MPD should always feed enough data into the pipe when playing local audio files, but for live streams this might happen. The `dryout_ms` parameter of the stream (default 2000ms) could reduce this problem.

**Comment by jgeisler0303** [comment]
Sample rate should be correct. I have `format "48000:16:2"` in `mpd.conf` `audio_output` section. But maybe mpd "thinks" 48kHz is slower than snapserver does?

Having watched the server and client output a little longer, I think the onResync happens more often than the audible hiccups and the client side `pShortBuffer->full() && (abs(shortMedian_) > 5)`. Are these two events caused by the same effect?

I think I upgraded Snapcast first after the problems started. But maybe there was some unrelated issue and what I see not is indeed due to the upgrade.

The problem occurs also with local files from mpd (or mplayer) but I'll try a higher `dryout_ms`. Or maybe I should give mpd a higher "nice" priority?

Is it worth to try the develop branch?

**Comment by jgeisler0303** [comment]
I didn't find the `dryout_ms` in the `snapserver.conf` is it an mpd parameter?

**Comment by badaix** [comment]
Yes, I highly recommend to test the develop branch, as the timing that is responsible for the buffer full messages is improved there.

**Comment by jgeisler0303** [comment]
I had `auto_resample "no"` set in `mpd.conf` it seems to be better now, but not gone.

In older Snapcast there used to be an argument `--streamBuffer` in the `/etc/default/snapserver`. Is that now the `dryout_ms` parameter?

**Comment by badaix** [comment]
`dryout_ms` seems to be a hidden (i.e. undocumented) feature :)
Just add it as parameter to your stream configuration in `snapserver.conf`:
```
[stream]
stream = pipe:///tmp/snapfifo?name=MPD&dryout_ms=5000
```
I checked the old code:
`--streamBuffer` is now this:
```
# Default stream read chunk size [ms]
#chunk_ms = 20
```

this is simply how much data the server should read in a row from the pipe. The default should be fine.

If you have a bad connection (I used to have a FRiTZ Box WiFi which I from time to time had to power-cycle, because I was facing the same problems you're describing), then it can help to change this parameter to a higher value:

```
# Buffer [ms]
#buffer = 1000
```

But this is just the overall record to playout latency, helping in bad wifi situations, but will not help getting rid of the `onResync` event.

I'm not aware of the `auto_resample` option, my config is looking like this:

```
audio_output {
    type            "fifo"
    name            "my pipe"
    path            "/tmp/snapfifo"
    format          "48000:16:2"
    mixer_type      "software"
}
```

**Comment by jgeisler0303** [comment]
OK, thanks. The higher `dry_out` doesn't change things. And neither does putting `Nice=-20` in the `/lib/systemd/system/mpd.servic`. My connection is Gigabit LAN. I have to do some other stuff now but will try the dev branch tonight.

**Comment by jgeisler0303** [comment]
Develop branch is compiling now.

Just out of curiosity I set `format 49000:16:2"` in mpd.conf audio_output. Now I have maybe a bit less `onResync` but it seems there are now hiccups without an accompanying `onResync`. Maybe now there is more data coming down the pipe than Snapcast is willing to handle. This leads me to the question: how does Snapcast actually make sure the fifo writer has the same idea of the data rate? I mean, if there is even the slightest difference in how the applications calculate the 48kHz from the system clock, there will be a drift which will either lead to the fifo running out of data or constantly growing in buffer size. There is no flow control going on between Snapcast and the fifo-feeder, right?

Is this maybe a fundamental flaw? Would it maybe make sense to implement an ALSA sink in Snapcast to have a single source of flow-rate (ALSA) that is communicated to both the reader and the writer? I imagine ALSA works something like this: is tell the writer: "you have to supply another buffer **now**" and the reader: "here is your next buffer, take it!".

**Comment by badaix** [comment]
I'm using MPD since the beginning of Snapcast w/o any issues. 
MPD writes the data as fast as it can to the fifo (having a capacity of 65k, when full, the fifo will block), and Snapcast is reading at 48kHz from the pipe.

**Comment by jgeisler0303** [comment]
Ah, ok. I didn't know there is a size limit and mpd fills is as fast as possible. That, of course, is a kind of flow control.

Maybe there is an Armbian specific problem with the fifo. At least in FreeBSD there is an issue like this: [https://forums.freebsd.org/threads/musicpd-skips-when-i-enable-a-fifo.63380/](https://forums.freebsd.org/threads/musicpd-skips-when-i-enable-a-fifo.63380/).

The develop branch didn't bring any improvement. But that was to be expected. I'll research some more into mpd and fifos now. Do you perhaps know any tools to debug the state of fifos?

**Comment by jgeisler0303** [comment]
Using [mbuffer](https://www.systutorials.com/docs/linux/man/1-mbuffer/) as a process stream input to snapserver I could now verify that mpd sends enough data. The buffer quickly filled to 100%. So the starving seems to happen at the reading end. A FIFO (named pipe) or process pipe should be basically the same.

`top`is showing ~35% CPU for mpd, 10% for snapserver and 15% for some kworker. That's already some load but the Banana Pi has two cores so its still far from 200%. Maybe next I'll have a look at the governor.

**Comment by badaix** [comment]
can you try with PCM codec (in `snapserver.conf`, section `[stream]` use `codec = pcm`).
You can also try to change the number of threads (e.g. `[server] threads = 4`).
There is a pool of worker threads. If all of them are busy with other things, like ogg encoding, there might be no thread available for reading new data from the pipe.

**Comment by jgeisler0303** [comment]
Hi, thanks for the suggestion. I had it already running on pcm but will try the increased number of threads.

Meanwhile I tried some things in the do_read function. I guess, I don't really understand the whole concept, but locally some things looked a bit strange to me, so I just changed them (I love open source!!!). Now I don't have the hard hiccups (onResync) any more and the smaller ones are much less frequent. This is how my do_read looks now:
```c++
void PosixStream::do_read()
{
    try
    {
        if (!stream_->is_open())
            throw SnapException("failed to open stream: \"" + uri_.path + "\"");

        int toRead = chunk_->payloadSize;
        int len = 0;
        do
        { 
            int count = read(stream_->native_handle(), chunk_->payload + len, toRead - len);
            if (count == 0)
            {
                throw SnapException("end of file");
            }
            if (count > 0)
            {
                len += count;
                bytes_read_ += len;
                idle_bytes_ = 0;
            }
        // just try to read till the time is up
        } while (len < toRead && (nextTick_ + buffer_ms_) < chronos::getTickCount());

        // if we didn't get all the needed data fill with zeros
        // I assume these zeros are really encoded, sent and played
        // so this is a way to "strech" the data if snapcast's idea of time is faster than that of the fifo feeder
        if (len < toRead)
        {
            memset(chunk_->payload + len, 0, toRead - len);
        }
        // I didn't really understand why we can wait longer the first time
        if (first_)
        {
            first_ = false;
            chronos::systemtimeofday(&tvEncodedChunk_);
            nextTick_ = chronos::getTickCount() + buffer_ms_;
        }

        encoder_->encode(chunk_.get());
        nextTick_ += chunk_ms_;
        long currentTick = chronos::getTickCount();

        // if we are late, insert a whole extra "stretch" buffer of zeros
        if (nextTick_ < currentTick)
        {
            memset(chunk_->payload, 0, toRead);
            encoder_->encode(chunk_.get());
            nextTick_ += chunk_ms_;
            currentTick = chronos::getTickCount();
        }       

        if (nextTick_ >= currentTick)
        {
            wait(read_timer_, std::chrono::milliseconds(nextTick_ - currentTick), [this] { do_read(); });
            return;
        }
        else
        {
            // if we are still late despite of the preceding "stretch" cycle, schedule the next do_read immediately
            wait(read_timer_, std::chrono::milliseconds(1), [this] { do_read(); });
            return;
        }

        lastException_ = "";
    }
    catch (const std::exception& e)
    {
        if (lastException_ != e.what())
        {
            LOG(ERROR, LOG_TAG) << "Exception: " << e.what() << std::endl;
            lastException_ = e.what();
        }
        disconnect();
        wait(read_timer_, 100ms, [this] { connect(); });
    }
}
```

Oh, and I realized that I am running MPD mostly with internet radio. That means that the radio station dictates the speed and amount of data. MPD cannot "stuff" the buffer as fast as it gets read because it cannot read the data arbitrarily. So, if the "real-time" radio has a slower notion of time than scnapcast (less data per second), the zero padding chunks are absolutely necessary. This seems to be the case for me, mostly. The other way around would mean there is more data than snapcast wants to read. In this case MPD purges the entire fifo by reading it itself to make room for more data (I checked this in the source code of MPD). If this happens, the hiccups are much worse because (in my case, check ``cat /proc/sys/fs/pipe-max-size``) 1MB of data or 5s time equivalent is lost (with the usual 65536 byte buffer it's still 350ms).

**Comment by badaix** [comment]
Nice, I will review your reader code tonight and compare it with the current code base.
I've refactored the threading model in v0.18.0, but didn't really change the inner loop code.
Regarding MPD: same for me, I'm using either librespot for Spotify or MPD for live radio (German WDR Stations, resampled by MPD to 48kHz), but strangely, I don't have problems with resyncs.

**Comment by jgeisler0303** [comment]
Maybe it has to do with the platform, i.e. the linux kernel and processor. I'm using a BananaPi which has two cores running at max 1GHz. Maybe the quality of the system clock is not so good. I wonder how the steady_timer manages to run without jumping while still not drift from the NTP time; or maybe it does? Maybe my system was always near the limit and now has a little more load.

Using four threads didn't really make a difference. What helps is a larger ``chunk_ms`` of 64ms. And what helps a lot is using a process stream via ``mbuffer -i /tmp/snapfifo -m 65536`` even though it add another process that need processor time. This setup also has the advantage that I can see how much is in the buffer. It showed me that even with the buffer filled the occasional failed reads occur. This leads me to believe that there might be some thing wrong with the piping of my kernel, because in theory pipes should be really fast.

BTW: I mostly listen to somafm and sending that directly to alsa is flawless.

I guess i'm quite happy the way it is now. But if it helps, I could do some more investigations. In any case, thanks for your help @badaix and the great work of Snapcast.

**Comment by badaix** [comment]
I've finished some refactoring work at the client. I checked your PosixStream modifications, and I think that the "nothing to read situation" (read < 0) cannot be ignored, also the reader shouldn't block, because it would block a thread from the limited thread pool, the whole server design is asynchronous.
I've introduced now a tolerance for the resync situation (resync happens when one read of a single chunk (e.g. 20ms PCM data) takes longer than the chunk's duration (e.g. 21ms will cause a resync of 1ms) and when many or some large resyncs happen, it might happen that the available overall buffer (default 1000ms) slowly vanishes.
Usually a single resync can be ignored, it will recover with the next read, this is what the tolerance will take care of. If it would sum up and exceed the tolerance, the resync event would reset the time stamps.
Also the silence handling is much improved from time stamping point of view.
Long story short: can you please the current develop branch? 
The current dev branch is supposed to be the v0.19 release candidate, I don't plan any new features, but will do some testing the next few days, as some crucial parts in the code base have changed.

**Comment by jgeisler0303** [comment]
Hi, I compiled the new code for client and server and the quality is much better! Thanks a lot for that.

Strangely though, I still get glitches when `(do_read) next read < 0 (process)` occurs (indeed, I think, there are even glitches without any apparent cause). Even more strangely, these glitches don't occur plus the `(do_read) next read < 0 (process)` occurs less frequently when I use `stream = process:///home/jgeisler/mb?name=process&log_stderr=true` where `/home/jgeisler/mb`is 

```
#!/bin/bash
mbuffer -i /tmp/snapfifo -m 65536
```

I guess, there is something fundamentally wrong in my Armbian with pipes or with the clock that boost uses.

But the level of hiccups / onResyncs is now bearable again and I can continue using SnapCast.

**Comment by badaix** [comment]
thanks, good to hear. Actually the "next read < 0" message shouldn't be audible, it's just a log message where formerly a resync happened. Also real resyncs are visible on the client side, in log lines like this:

```
2020-03-02 21-15-31 [Info] (Stream) Chunk: 0	0	0	0	500	119	1
2020-03-02 21-15-32 [Info] (Stream) Chunk: 0	0	0	0	500	119	0
2020-03-02 21-15-33 [Info] (Stream) pShortBuffer->full() && (abs(shortMedian_) > 5): -14603
2020-03-02 21-15-34 [Info] (Stream) Chunk: 0	0	0	0	500	118	0
```

**Comment by jgeisler0303** [comment]
Yes, I can see how `onResync` coincides with `pMiniBuffer->full() && (abs(pMiniBuffer->mean())` on the client side. Nonetheless, the glitches without `onResync`are real when I use the direct `stream = pipe` method. But I think, the cause at the mpd-end of the pipe because there is another method where the glitches don't occur. This method is to configure mpd to output to alsa default and to put the following in `/var/lib/mpd/.asoundrc`

```
pcm.!default {
    type plug
    slave {
        pcm rate48000Hz # Direct default output to the below converter
    }
}

pcm.rate48000Hz {
    type rate
    slave {
        pcm writeFile # Direct to the plugin which will write to a file
        format S16_LE
            #        channels 2
            rate 48000
    }
    #route_policy copy
}

pcm.writeFile {
    type file
    slave {
        pcm "hw:0,0" # Now write to the actual sound card
    }
}
```

But I think I'll go with the "mbuffer" solution, and maybe invest in a raspi4 to match the growing demands of my server :-)

**Comment by badaix** [comment]
fixed resyncs in v0.19.0
